File: java/src/ch1/Ex122.java
package ch1;

import java.util.Date;

public class Ex122 {
    public static void main(String[] args) {
        System.out.println(runtime());
        searchForPrimes(1000, 1020);
        searchForPrimes(10000, 10040);
        searchForPrimes(100000, 100080);
        System.out.println("done");
    }

    public static boolean odd(int n) {
        return 1 == n % 2;
    }

    public static int square(int n) {
        return n * n;
    }

    public static boolean divides(int a, int b) {
        return 0 == b % a;
    }

    public static boolean prime(int n) {
        return new SmallestDivisor(n).value() == n;
    }

    public static long runtime() {
        return new Date().getTime();
    }

    public static void reportPrime(long elapsedTime) {
        System.out.print(" *** ");
        System.out.println(elapsedTime);;
    }

    public static void reportNonPrime() {
        //do nothing System.out.println("  ");
    }

    public static void startPrimeTest(int n, long startTime) {
        if (prime(n)) {
            reportPrime(runtime() - startTime);
        } else {
            reportNonPrime();
        }
    }

    public static void timedPrimeTest(int n) {
        System.out.println();
        System.out.print(n);
        startPrimeTest(n, runtime());
    }

    public static void searchForPrimes(int start, int end) {
        searchHelper(odd(start) ? start : ++start, end);
        System.out.println("finished");
    }

    public static void searchHelper(int n, int end) {
        if (n <= end) {
            timedPrimeTest(n);
            searchHelper(n+2, end);
        } else {
            System.out.println(" ");
        }
    }

    static class SmallestDivisor implements Value<Integer> {
        int n;
        SmallestDivisor(int number) {
            n = number;
        }

        @Override
        public Integer value() {
            return findDivisor(n, 2);
        }

        int findDivisor(int n, int test_divisor) {
            if (square(test_divisor) > n) {
                return n;
            } else if (divides(test_divisor, n)) {
                return test_divisor;
            } else {
                return findDivisor(n, test_divisor + 1);
            }
        }
    }
}

File: elixir/ch1/ex122.ex
# cd("c:/Users/Yohan/IdeaProjects/sicp/elixir/ch1")
# Load with import_file("filename")
# Run with Module.func(args)
defmodule Ex122 do
  @moduledoc false

  def sq(x), do: x * x

  def divides?(a, b), do: rem(b, a) == 0

  def smallest_divisor(n), do: find_divisor(n, 2)

  def find_divisor(n, test_divisor) when test_divisor * test_divisor > n, do: n
  def find_divisor(n, test_divisor) do
    if divides?(test_divisor, n) do
      test_divisor
    else
      find_divisor(n, test_divisor + 1)
    end
  end

  def prime?(n), do: n == smallest_divisor(n)

  def prime_test(n) do
    if prime?(n) do
      IO.puts("#{n} ***")
    else
      IO.puts("#{n}")
    end
  end

  def search_helper(n, finish) when n <= finish do
    prime_test(n)
    search_helper(n + 2, finish)
  end
  def search_helper(_, _), do: IO.puts("finished")

  def search_for_primes(start, finish) when rem(start, 2) == 1, do: search_helper(start, finish)
  def search_for_primes(start, finish), do: search_helper(start + 1, finish)
end
